nomad/plan_apply_node_tracker.go

```diff
@@ -0,0 +1,203 @@
+// the time window.
+func (c *CachedBadNodeTracker) IsBad(nodeID string) bool {
+	logger := c.logger.With("node_id", nodeID)
+	logger.Debug("checking if node is bad")
+
+	value, ok := c.cache.Get(nodeID)
+	if !ok {
+		logger.Debug("node not in cache")
+		return false
+	}
+
+	// Limit the number of nodes we report as bad to avoid mass assigning nodes
+	// as ineligible, but do it after Get to keep the cache entry fresh.
+	if !c.limiter.Allow() {
+		logger.Info("returning false due to rate limiting")
+		return false
+	}
+
+	stats := value.(*badNodeStats)
+	score := stats.score()
+
+	if score >= c.threshold {
+		logger.Debug("node is bad", "score", score)
+		return true
+	}
+
+	logger.Debug("node is not bad", "score", score)
+	return false
+}
+
+// Add records a new rejection for node. If it's the first time a node is added
+// it will be included in the internal cache. If the cache is full the least
+// recently updated or accessed node is evicted.
+func (c *CachedBadNodeTracker) Add(nodeID string) {
+	value, ok := c.cache.Get(nodeID)
+	if !ok {
+		value = newBadNodeStats(c.window)
+		c.cache.Add(nodeID, value)
+	}
+
+	stats := value.(*badNodeStats)
+	score := stats.record()
+	c.logger.Debug("adding node plan rejection", "node_id", nodeID, "score", score)
+}
+
+// EmitStats generates metrics for the bad nodes being currently tracked. Must
+// be called in a goroutine.
+func (c *CachedBadNodeTracker) EmitStats(period time.Duration, stopCh <-chan struct{}) {
+	timer, stop := helper.NewSafeTimer(period)
+	defer stop()
+
+	for {
+		timer.Reset(period)
+
+		select {
+		case <-timer.C:
+			c.emitStats()
+		case <-stopCh:
+			return
+		}
+	}
+}
+
+func (c *CachedBadNodeTracker) emitStats() {
+	for _, k := range c.cache.Keys() {
+		value, _ := c.cache.Get(k)
+		stats := value.(*badNodeStats)
+		score := stats.score()
+
+		labels := []metrics.Label{
+			{Name: "node_id", Value: k.(string)},
+		}
+		metrics.SetGaugeWithLabels([]string{"nomad", "plan", "rejection_tracker", "node_score"}, float32(score), labels)
+	}
+}
+
+// badNodeStats represents a node being tracked by a BadNodeTracker.
+type badNodeStats struct {
+	history []time.Time
+	window  time.Duration
+}
+
+// newBadNodeStats returns an empty badNodeStats.
+func newBadNodeStats(window time.Duration) *badNodeStats {
+	return &badNodeStats{
+		window: window,
+	}
+}
+
+// score returns the number of rejections within the past time window.
+func (s *badNodeStats) score() int {
+	count := 0
+	windowStart := time.Now().Add(-s.window)
+
+	for i := len(s.history) - 1; i >= 0; i-- {
+		ts := s.history[i]
+		if ts.Before(windowStart) {
+			// Since we start from the end of the history list, anything past
+			// this point will have happened before the time window.
+			break
+		}
+		count += 1
+	}
+	return count
+}
+
+// record adds a new entry to the stats history and returns the new score.
+func (s *badNodeStats) record() int {
+	now := time.Now()
+	s.history = append(s.history, now)
+	return s.score()
+}
```

Comment:
On large clusters I'd expect the LRU would age-out most of the tracked nodes eventually. But I think there's a pathological case here where a stable cluster with 50 or fewer nodes will continue to grow the size of the `badNodeStats.history` forever, so long as none of the nodes get plan rejections more frequently than the window allows. 

It's a small bit of data (24 bytes each rejection), but we could eliminate it entirely if we dropped the old entries either when we called `record()` or when we called `score()`.

Reply:
Good catch! I made it so `score` will remove entries that happened before the window start, so they will never be considered any more.